Fix requests with large bodies - fixes #220

mruoss opened the pull request:
As raised in issue #220 we currently have a problem with large resources / request bodies. 

- [x] Entry in CHANGELOG.md was created
- [ ] New unit tests were created
- [x] New integration tests were created

mruoss committed:
add failing test

mruoss commented:
@elliottneilclark FYI.

As mentioned in the issue #220, I've started by adding a failing test.

mruoss committed:
set empty body to nil

mruoss committed:
stream request body to server

mruoss commented:
I can't figure out how to make dialyzer happy with this code (see https://github.com/elixir-mint/mint/issues/380). Besides that, I think this should do the trick. @elliottneilclark would you care to test your code using this branch?

mruoss committed:
remove wip tag

mruoss committed:
add changelog

mruoss committed:
make the dialyzer happy. (oh my god)

mruoss commented:
Dialyzer is happy now. Ready to review.

elliottneilclark commented on lib/k8s/client/mint/dialyzer_hacks.ex:
```diff
@@ -0,0 +1,11 @@
+  """
+
+  @doc """
+  This function tells Dialyzer that the conn is actually a HTTP2 conn.
```
Might want the why or the story of how this came to be, those questions are important but often lost to time.

mruoss replied:
Indeed, thanks.

elliottneilclark commented on lib/k8s/client/mint/request.ex:
```diff
@@ -155,4 +174,46 @@ defmodule K8s.Client.Mint.Request do
+          Enum.split(pending_request_body, chunk_size)
+      end
+
+    with {:ok, conn} <- Mint.HTTP.stream_request_body(conn, request_ref, chunk),
```
Chunk here is not always the expected types of iodata. https://github.com/elixir-mint/mint/blob/v1.4.2/lib/mint/http2.ex#L508

When starting a watch on just about every kind on a cluster I'm seeing:
```
  * 1st argument: not an iodata term                                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                                               
            :erlang.iolist_size([123, 34, 97, 112, 105, 86, 101, 114, 115, 105, 111, 110, 34, 58, 34, 97, 112, 105, 101, 120, 116, 101, 110, 115, 105, 111, 110, 115, 46, 107, 56, 115, 46, 105, 111, 47, 118, 49, 34, 44, 34, 107, 105, 110, 100, 34, 58, 34, 67, 117, ...])                                  
            (mint 1.4.2) lib/mint/http2.ex:1196: Mint.HTTP2.encode_data/4                                                                                                                                                                                                                                      
            (mint 1.4.2) lib/mint/http2.ex:526: Mint.HTTP2.stream_request_body/3                                                                                                                                                                                                                               
            (k8s 2.0.0) lib/k8s/client/mint/request.ex:196: K8s.Client.Mint.Request.stream_request_body/2                                                                                                                                                                                                      
            (k8s 2.0.0) lib/k8s/client/mint/http_adapter.ex:171: K8s.Client.Mint.HTTPAdapter.handle_call/3                                                                                                                                                                                                     
            (stdlib 4.2) gen_server.erl:1149: :gen_server.try_handle_call/4                                                                                                                                                                                                                                    
            (stdlib 4.2) gen_server.erl:1178: :gen_server.handle_msg/6                                                                                                                                                                                                                                         
            (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3                                                                                                                                                                                                                                         
    (elixir 1.14.2) lib/gen_server.ex:1038: GenServer.call/3                                                                                                                                                                                                                                                   
    (k8s 2.0.0) lib/k8s/client/mint_http_provider.ex:41: anonymous fn/3 in K8s.Client.MintHTTPProvider.stream/5                                                                                                                                                                                                
    (elixir 1.14.2) lib/stream.ex:1623: Stream.do_resource/5                                                                                                                                                                                                                                                   
    (elixir 1.14.2) lib/stream.ex:942: Stream.do_transform/5                                                                                                                                                                                                                                                   
    (elixir 1.14.2) lib/stream.ex:1811: Enumerable.Stream.do_each/4                                                                                                                                                                                                                                            
    (elixir 1.14.2) lib/stream.ex:942: Stream.do_transform/5                                                                                                                                                                                                                                                   
    (elixir 1.14.2) lib/enum.ex:4307: Enum.each/2                   
```

mruoss replied:
Oh boy yeah that's me being ignorant about iodata / binary / string / charlist.

elliottneilclark commented on lib/k8s/client/mint/http_adapter.ex:
```diff
@@ -148,20 +148,30 @@ defmodule K8s.Client.Mint.HTTPAdapter do
+    {body, pending_request_body} =
+      case body do
+        nil -> {body, body}
+        body -> {:stream, String.to_charlist(body)}
```
We don't need to stream if the entire request fits in the window. I'm pretty sure that saves a round trip, so might be a nice optimization to put as a TODO.

mruoss replied:
Indeed, nor if it's HTTP1. 
The HTTP1 case I'll do now. The other one probably a todo for now.

mruoss replied:
actually, not I can add both checks - non need for the TODO

elliottneilclark commented:
This looks super close, thank you! I'm seeing an issue after starting a watch and streaming the events.

mruoss committed:
break test again by using utf-8 icon in resource

mruoss committed:
allow TODOs in the code

mruoss committed:
use binary operations to create chunks

mruoss commented:
> This looks super close, thank you! I'm seeing an issue after starting a watch and streaming the events.

Oh boy yeah that's me being ignorant about iodata / binary / string / charlist. Added an emoticon to the annotation => test fails. But fixed it now. Together with the other feedback. Thanks for the review!

elliottneilclark committed:
test: add on kind integration testing in addition to k3d

Kind is another way of running kubernetes. https://kind.sigs.k8s.io/
Since kind is packaged and released by the CNCF and Kubernetes teams
it tends to be very close to trunk and bleeding edge.

This commit adds kd3 and kind commands to the makefile. Then it wires
the test command to run kind and k3d testing sequentially. This shows
the differences (including the streaming body size issue).

mruoss commented:
Opened https://github.com/elixir-mint/mint/issues/381

mruoss committed:
let's always stream the request body for HTTP2 connections

mruoss committed:
optimize Makefile to work with different config files for Kind and K3d

Makefile detects whether a file is there and does not recreate the cluster if the config file is there. Also, make test is meant to run ALL the tests, not just integration tests.

mruoss committed:
remove DialyzerHacks module and integrate get_window_size/2 in Request module

elliottneilclark reviewed (approved):
Thank you!
